doc/ql.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd June 14, 2009
.Os "Mac OS X"
.Dt QL \&1 "CLIMac Reference Manual"
.Sh NAME                 \" Section Header - required - don't modify 
.Nm ql
.Nd Generate Quick Look thumbnail or preview
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Fl p
.Ar file
.Op Fl o Ar out
.Nm
.Fl t
.Op Fl w Ar width Fl h Ar height
.Op Fl s Ar scale
.Op Fl o Ar out.tiff
.Sh DESCRIPTION          \" Section Header - required - don't modify
.Pp
In the first synopsis form, generates a QuickLook preview of the given file. The type of the generated preview as a UTI (eg. public.html) is printed. In the second synopsis form, generates a thumbnail image in TIFF format.
.Pp
Note that not all files can be previewed with Quick Look -- this is dependent on what Quick Look generators you have installed.
.\".Sh FILES                \" File used or created by the topic of the man page
.Sh EXAMPLES
.Ss Thumbnail
.Pp
Generate a 512 pixel high and 512 pixel wide thumbnail of the file at
.Pa /usr/include/libc.h
and save the image as
.Pa libc.tiff :
.Pp
.Dl ql -t /usr/include/libc.h -h 512 -w 512 -o libc.tiff
.Pp
As above, but pipe to
.Xr imgconv 1
to generate a PNG image:
.Pp
.Dl ql -t /usr/include/libc.h -h 512 -w 512 | imgconv libc.png
.Ss Preview
.Pp
Generate a preview of the specified file. The preview can be an image, audio, PDF, HTML or other text format.
.Pp
.Dl ql -p ~/Documents/WWDC.key -o output
.Sh DIAGNOSTICS
The
.Nm
utility exits 0 on success, and \*(Gt0 if an error occurs.
.\".Sh COMPATIBILITY
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr imgconv 1 ,
.Xr qlmanage 1
.\" .Sh BUGS              \" Document known, unremedied bugs
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner
.Sh AUTHORS
.Pp
The
.Nm
utility was written by
.An Sq Vacuous Virtuoso
.Aq http://lipidity.com

src/ql.m
#import <Cocoa/Cocoa.h>
#import <QuickLook/QuickLook.h>

typedef void *QLPreviewRef;
extern QLPreviewRef QLPreviewCreate(void *unknownNULL, CFTypeRef item, CFDictionaryRef options);
extern CFDataRef QLPreviewCopyData(QLPreviewRef thumbnail);
extern CFURLRef QLPreviewCopyURLRepresentation(QLPreviewRef);
extern CFDictionaryRef QLPreviewCopyOptions(QLPreviewRef);
extern CFDictionaryRef QLPreviewCopyProperties(QLPreviewRef);
extern CFStringRef QLPreviewGetPreviewType(QLPreviewRef); // eg. public.webcontent; public.text; public.image; public.pdf
extern void QLPreviewSetPreviewType(QLPreviewRef, CFStringRef);
extern void QLPreviewSetForceContentTypeUTI(QLPreviewRef, CFStringRef);

typedef void *QLThumbnailRef;
#if 0
extern const NSString *kQLThumbnailOptionContentTypeUTI;
//extern const NSString *kQLThumbnailOptionIconModeKey;
extern QLThumbnailRef QLThumbnailCreate(void *unknownNULL, CFURLRef fileURL, CGSize iconSize, CFDictionaryRef options);
extern CGImageRef QLThumbnailCopyImage(QLThumbnailRef thumbnail);
_QLThumbnailSupportsContentUTIAtSize
_QLThumbnailCopySpecialGenericImage
_QLThumbnailGetMaximumSize
_QLThumbnailGetMinimumUsefulSize
_QLThumbnailSetForceContentTypeUTI
#endif

int main (int argc, char *argv[]) {
	if (argc > 1) {
		NSAutoreleasePool *pool = [NSAutoreleasePool new];
		const struct option longopts[] = {
			// actions
			{ "preview", no_argument, NULL, 'p' },
			{ "thumbnail", no_argument, NULL, 't' },
			// thumbnail options
			{ "width", required_argument, NULL, 'w' },
			{ "height", required_argument, NULL, 'h' },
			{ "icon-mode", no_argument, NULL, 'm' },
			{ "scale", required_argument, NULL, 's' },
			// output
			{ "output", required_argument, NULL, 'o' },
			{ NULL, 0, NULL, 0 }
		};
		int c;
		char action = 0;
		NSURL *oURL = nil;
		CGSize size = {128.0f, 128.0f};
		float scale = 0.0f;
		BOOL icon = NO;
		while ((c = getopt_long_only(argc, (char **)argv, "pto:w:h:s:m", longopts, NULL)) != EOF) {
			switch (c) {
				case 't':
				case 'p':
					if (action == 0)
						action = c;
					else
						errx(1, "Only one %s option may be specified", "-thumbnail or -preview");
					break;
				case 'o':
					if (oURL == nil)
						oURL = (NSURL *)CFURLCreateFromFileSystemRepresentation(NULL, (UInt8 *)optarg, strlen(optarg), false);
					else
						errx(1, "Only one %s option may be specified", "-output");
					break;
				case 'h': {
					char *ptr;
					size.height = strtof(optarg, &ptr);
					if (ptr == optarg)
						errx(1, "Invalid argument to %s option", "-height");
					break;
				}
				case 'w': {
					char *ptr;
					size.width = strtof(optarg, &ptr);
					if (ptr == optarg)
						errx(1, "Invalid argument to %s option", "-width");
					break;
				}
				case 's': {
					char *ptr;
					scale = strtof(optarg, &ptr);
					if (ptr == optarg)
						errx(1, "Invalid argument to %s option", "-scale");
					break;
				}
				case 'm':
					icon ^= YES;
					break;
				default: goto usage;
			}
		}
		if (action == 0)
			errx(1, "One of -preview of -thumbnail must be specified");
		if (oURL == nil && isatty(STDOUT_FILENO))
			errx(1, "Refusing to dump data to a terminal");
		argc -= optind;
		if (argc == 1) {
			argv += optind;
			NSData *tiff = nil;
			CFTypeRef item = NULL;
			if (action == 'p' && argv[0][0] == '-' && argv[0][1] == '\0') {
				item = (CFDataRef)[[NSFileHandle fileHandleWithStandardInput] readDataToEndOfFile];
			} else {
				item = CFURLCreateFromFileSystemRepresentation(NULL, (UInt8 *)argv[0], strlen(argv[0]), false);
			}
			if (action == 'p') {
				QLPreviewRef ql = QLPreviewCreate(NULL, item, NULL);
				tiff = (NSData *)QLPreviewCopyData(ql);
				if (tiff != nil)
					fprintf(stderr, "Generated Preview: %s\n", [(NSString *)QLPreviewGetPreviewType(ql) fileSystemRepresentation]);
			} else {
				NSMutableDictionary *opts = [NSMutableDictionary dictionary];
				if (scale)
					[opts setObject:[NSNumber numberWithFloat:scale] forKey:(NSString *)kQLThumbnailOptionScaleFactorKey];
				if (icon)
					[opts setObject:[NSNumber numberWithBool:YES] forKey:(NSString *)kQLThumbnailOptionIconModeKey];
				CGImageRef img = QLThumbnailImageCreate(NULL, item, size, (CFDictionaryRef)opts);
				if (img != NULL) {
					NSBitmapImageRep *bitmap = [[NSBitmapImageRep alloc] initWithCGImage:img];
					tiff = [[bitmap TIFFRepresentationUsingCompression:NSTIFFCompressionLZW factor:0.0f] retain];
					[bitmap release];
					CGImageRelease(img);
				}
			}
			if (tiff != nil) {
				if (oURL != nil) {
					NSError *error = [NSError errorWithDomain:NSPOSIXErrorDomain code:0 userInfo:nil];
					if (![tiff writeToURL:oURL options:0U error:&error])
						errx(1, "Write failed: %s", [[error localizedFailureReason] fileSystemRepresentation]);
				} else {
					if (fwrite([tiff bytes], [tiff length], 1, stdout) != 1)
						err(1, NULL);
				}
				return 0;
			}
		} else
			goto usage;
		[pool release];
		return 1;
	}
usage:
	fprintf(stderr, "usage:  %s -p <file> [-o <out>]\n\t%s -t [-w <width>] [-h <height>] [-s <scale>] <file> [-o <out>.tiff]\n", argv[0], argv[0]);
	return 1;
}
